c2rust-transpile/src/c_ast/mod.rs

```diff
@@ -679,6 +750,103 @@ impl TypedAstContext {
+    /// newer; on these later Clang versions, it should have no effect.
+    ///
+    /// This pass is necessary because we reify some typedef types (such as `size_t`) into their own
+    /// distinct Rust types. As such, we need to make sure we know the exact type to generate when
+    /// we translate an expr, not just its resolved type (looking through typedefs).
+    pub fn bubble_expr_types(&mut self) {
+        struct BubbleExprTypes<'a> {
+            ast_context: &'a mut TypedAstContext,
+        }
+
+        use iterators::{immediate_children_all_types, NodeVisitor};
+        impl<'a> NodeVisitor for BubbleExprTypes<'a> {
+            fn children(&mut self, id: SomeId) -> Vec<SomeId> {
+                immediate_children_all_types(self.ast_context, id)
+            }
+            fn post(&mut self, id: SomeId) {
+                if let SomeId::Expr(e) = id {
+                    let new_ty = match self.ast_context.c_exprs[&e].kind {
+                        CExprKind::Conditional(_ty, _cond, lhs, rhs) => {
+                            let lhs_type_id =
+                                self.ast_context.c_exprs[&lhs].kind.get_qual_type().unwrap();
+                            let rhs_type_id =
+                                self.ast_context.c_exprs[&rhs].kind.get_qual_type().unwrap();
+
+                            let lhs_resolved = self.ast_context.resolve_type_id(lhs_type_id.ctype);
+                            let rhs_resolved = self.ast_context.resolve_type_id(rhs_type_id.ctype);
+
+                            if CTypeKind::PULLBACK_KINDS
+                                .contains(&self.ast_context[lhs_resolved].kind)
+                            {
+                                Some(lhs_type_id)
+                            } else if CTypeKind::PULLBACK_KINDS
+                                .contains(&self.ast_context[rhs_resolved].kind)
+                            {
+                                Some(rhs_type_id)
+                            } else {
+                                None
+                            }
+                        }
+                        CExprKind::Binary(_ty, op, lhs, rhs, _, _) => {
+                            let rhs_type_id =
+                                self.ast_context.c_exprs[&rhs].kind.get_qual_type().unwrap();
+                            let lhs_kind = &self.ast_context.c_exprs[&lhs].kind;
+                            let lhs_type_id = lhs_kind.get_qual_type().unwrap();
+
+                            let lhs_resolved = self.ast_context.resolve_type_id(lhs_type_id.ctype);
+                            let rhs_resolved = self.ast_context.resolve_type_id(rhs_type_id.ctype);
+
+                            let neither_ptr = !self.ast_context[lhs_resolved].kind.is_pointer()
+                                && !self.ast_context[rhs_resolved].kind.is_pointer();
+
+                            if op.all_types_same() && neither_ptr {
```

Comment:
Could you apply integer promotion rules here to cover more combinations? E.g. a `uint16_t + uint32_t`  you know will be  a `uint32_t`.

You could leave it for later with a TODO comment.

Reply:
The Clang AST does integer promotion explicitly so that we don't have to worry about these cases.